File: 2020/two/twopointtwo.py
#!/user/bin/env python3

from sys import argv

def parse(filename):
  db = []
  with open(filename, 'r') as infile:
    for line in infile:
      pol = {}
      policy,password = line.split(":")
      pol["min"] = policy.split(" ")[0].split("-")[0]
      pol["max"] = policy.split(" ")[0].split("-")[1]
      pol["chr"] = policy.split(" ")[1]
      pol["pwd"] = password.strip()
      db.append(pol)
  return (db)

def check(policy):
  first = int(policy["min"])-1 #not zero indexed
  second = int(policy["max"])-1
  char = policy["chr"]
  password = policy["pwd"]
  if password[first] != char and \
     password[second] != char:
    return (1) #neither position matches
  if password[first] == char and \
     password[second] == char:
    return (2) #both position matches
  if password[first] == char or \
     password[second] == char:
    return (0) #only one position matches
  exit(1) #Something went wrong


if __name__ == '__main__':
  db = parse(argv[1])
  for policy in db:
    result = check(policy)
    if result != 0:
      print(f'[*] Password "{policy["pwd"]}" violated policy "{policy["min"]}-{policy["max"]} {policy["chr"]}"')
      if result == 1:
        print('  - Neither character matches')
      if result == 2:
        print('  - Both characters match')
    else:
      print(f'[+] Password "{policy["pwd"]}" meets policy "{policy["min"]}-{policy["max"]} {policy["chr"]}"')

File: 2020/two/twopointtwo.rs
use std::env;
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where P: AsRef<Path>, {
  let file = File::open(filename)?;
  Ok(io::BufReader::new(file).lines())
}

#[derive(Debug)]
struct Policy {
  min: i32,
  max: i32,
  chr: String,
  pwd: String,
}

fn parse(filename: &String) -> Vec<Policy>{
  let mut db: Vec<Policy> = Vec::new();
  if let Ok(lines) = read_lines(filename) {
    for line in lines {
      if let Ok(result) = line {
        let (mut count, mut min, mut max, mut chr, mut pwd) = (0, "", "", "", "");
        for s in result.split(&['-', ' ', ':'][..]) {
          match count {
            0 => min = s,
            1 => max = s,
            2 => chr = s,
            4 => pwd = s,
            _ => (),
          }
          count += 1;
        }
        let pol: Policy = Policy {
          min: min.parse().unwrap_or(-1),
          max: max.parse().unwrap_or(-1),
          chr: chr.to_string(),
          pwd: pwd.trim().to_string()
        };
        db.push(pol);
      }
    }
  }
  return db;
}

fn check(pol: &Policy) -> Result<usize,&str>
{
  if pol.max == -1 || pol.min == -1 {
    return Err("Error parsing policy.");
  }
  let first = pol.pwd.get(pol.min as usize-1..pol.min as usize).unwrap_or("xx");
  let second = pol.pwd.get(pol.max as usize-1..pol.max as usize).unwrap_or("xx");
  if first == "xx" || second == "xx" {
    return Err("Error parsing password.");
  }
  if first != pol.chr && second != pol.chr {
    return Err("Neither character matches.");
  } else if first == pol.chr && second == pol.chr {
    return Err("Both positions match.");
  } else if first == pol.chr || second == pol.chr {
    return Ok(1)
  }
  Err("Something went wrong")
}

fn main() {
  let args: Vec<String> = env::args().collect();
  let db = parse(&args[1]);
  for record in db {
    match check(&record) {
      Err(e) => println!("[*] Password {} violated policy {}-{} {}: {}", record.pwd, record.min, record.max, record.chr, e),
      Ok(_) => println!("[+] Password {} meets policy {}-{} {}", record.pwd, record.min, record.max, record.chr), 
    }
  }
}
